### ISSUE TITLE
Enforce a specific ordering of attributes

### ISSUE THREAD

**Issue by jdforrester** [wikimedia/stylelint-config-wikimedia#2]
```
"declaration-block-properties-order": [ "content", "background-image", "backg
round-origin", "background-position", "background-repeat", "background-size", "color"
, "list-style", "filter", "opacity", "display", "clip", "overflow", "visibility", "fl
oat", "clear", "position", "top", "left", "bottom", "right", "z-index", "box-sizing",
 "min-width", "max-width", "width", "height", "outline", "margin", "padding", "border
", "border-radius", "box-shadow", "direction", "hyphens", "font-family", "font-size",
 "font-weight", "line-height", "text-align", "text-decoration", "text-overflow", "tex
t-shadow", "text-transform", "vertical-align", "white-space", "cursor", "zoom", "anim
ation", "transition" ],
```

**Comment by Nikerabbit** [comment]
Where does this order come from? Csscomb has some presets: https://github.com/csscomb/csscomb.js/tree/master/config

**Comment by edg2s** [comment]
...or don't do this. It has little benefit (selector lists are usually not that long so keeping them ordered doesn't make them more readable), and will just waste developer time fixing more "errors".

**Comment by Nikerabbit** [comment]
If the fixing can be automated, I don't see it as a problem. The benefits are not that big for well written CSS because this does not really affect inside-a-block readability, but for less than well written CSS it is helpful, because it makes comparing separate blocks easier.

**Comment by Volker-E** [comment]
We've had some conversation about that already on the [CSS Coding Conventions talk page](https://www.mediawiki.org/wiki/Manual_talk:Coding_conventions/CSS#Order_of_declarations) including a proposal from my years of working with CSS.
Ability to scan over selectors quickly and find the most important (most often used) properties, like colors, display, position easily or orientate in box model definition by not changing order of (from outside to inside) margin, padding and border constantly from one selector to another are bringing –when working in an environment like ours– strong advantages in my opinion.

**Comment by edg2s** [comment]
**If** the fixing can be automated it is less objectionable - although I maintain it is exceptional for a class to have more than rules that can be scanned easily, so it still seems more trouble that is it worth.

**Comment by Nikerabbit** [comment]
csscomb can automate it, but I have no experience on stylelint.

**Comment by Volker-E** [comment]
I agree, that not-automated ordering might seem like a burden or even out of scope. Maybe it would be preferable as a build step after developer has ended work.
For completely new projects it would make sense though to have a guideline (and a possible enforcement). As I'm spending a good amount of time in CSS/Less(/SCSS/Post-processor enhanced) files, I'm personally convinced, that certain property order is a time-saver.

**Comment by edg2s** [comment]
Users may also want to group non adjacent rules for clarity, consider:
// Reduce width for border
width: 98px;
border: 2px;

**Comment by jdforrester** [comment]
(Upstream has deprecated this from core and moved into a plugin.)

**Comment by Volker-E** [comment]
With ['stylelint-order'](https://github.com/hudochenkov/stylelint-order) a maintained and powerful stylelint plugin existst. 
Provided https://gerrit.wikimedia.org/r/#/c/oojs/ui/+/483638/ as example. It has stylelint autofix capabilities, although the example given by @edg2s above shows, why autofix should be used with a grain of salt.
As somebody who spends a good time in CSS, I'd be still inclined to have some order of properties for cleaner, better readable CSS structure.

**Comment by edg2s** [comment]
Per discussion this might be useful an optional plugin (e.g. `wikimedia/ordered`) that could be used by CSS-heavy projects such as OOUI. I would suggest just using custom rules in OOUI to begin with, then we can extract to a plugin if there is demand elsewhere.

**Comment by danisztls** [comment]
csscomb is not maintained since 2019

Alternatives:
- https://github.com/Siilwyn/css-declaration-sorter/
- https://github.com/hudochenkov/postcss-sorting/

**Comment by Volker-E** [comment]
Thanks @danisztls. Note that we've already added and are actively using stylelint-order to [Codex](https://gerrit.wikimedia.org/r/admin/repos/design/codex,general) since my last comment here.
